common.rkt

```diff
@@ -23,50 +25,152 @@
 
 ;; States
 (define empty-sub '())
+(define empty-diseq '())
+(define empty-types '())
+
 (define (walk t sub)
   (let ((xt (and (var? t) (assf (lambda (x) (var=? t x)) sub))))
     (if xt (walk (cdr xt) sub) t)))
+
 (define (occurs? x t sub)
   (cond ((pair? t) (or (occurs? x (walk (car t) sub) sub)
                        (occurs? x (walk (cdr t) sub) sub)))
         ((var? t)  (var=? x t))
         (else      #f)))
+
+(define (walk-types t types)
+  (cond
+    ((number? t) number?)
+    ((symbol? t) symbol?)
+    ((string? t) string?)
+    ((var? t) (let* ((xt (assf (lambda (x) (var=? t x)) types)))
+                (if xt (cdr xt) #f)))
+    (else #f)))
+
 (define (extend-sub x t sub)
   (and (not (occurs? x t sub)) `((,x . ,t) . ,sub)))
 
-(struct state (sub) #:prefab)
-(define empty-state (state empty-sub))
+(define (extend-diseq =/=s diseq)
+  (cons =/=s diseq))
+
+(define (extend-types x t types)
+  `((,x . ,t) . ,types))
+
+(define (reduce-types t types)
+  (filter (lambda (type-constraint) (not (eq? t (car type-constraint)))) types))
```

Comment:
I would call this `var-type-remove`.  `t` is always a variable.

You can also use the `remove` procedure (providing a third argument) instead of `filter` for this since there is at most one type entry for this variable.  See: https://docs.racket-lang.org/reference/pairs.html#%28def._%28%28lib._racket%2Fprivate%2Flist..rkt%29._remove%29%29